app/discovery/discovery.go

```diff
@@ -146,6 +148,37 @@ func (s *Service) Match(srv, src string) (string, MatchType, bool) {
 	return src, MTProxy, false
 }
 
+// ScheduleHealthCheck starts background loop with health-check
+func (s *Service) ScheduleHealthCheck(ctx context.Context, interval time.Duration) {
+	log.Printf("health-check scheduled every %s", interval)
+
+	go func() {
+	hloop:
+		for {
+			timer := time.NewTimer(interval)
+			select {
+			case <-timer.C:
+				s.lock.RLock()
+				cres := CheckHealth(s.Mappers())
+				s.lock.RUnlock()
+
+				// alive services would be picked up first
```

Comment:
not sure what the reason for this sort

Reply:
As you mentioned, we want that the alive destination will be picked up if multiple destinations defined for the same route. After sort I have all alive destination on the top and dead in the and of the slice. And if I correctly understood method [Match](https://github.com/umputun/reproxy/blob/a73a16c12bbd376859667cb42258206a595d6c92/app/discovery/discovery.go#L130), the destination would be picked up exactly in the order they were inserted.

Reply:
two things:

- generally speaking, resorting is a dangerous idea here as the original sort important. I have not checked if you preserved the order, but even if this the case such approach may lead to some unexpected results, for example, you will put dead matcher to the end and it may trigger some rule above unrelated to the given route, for example, some "catch-all" for static assets
- selection of one of the conflicting routes is a use case for sure, but a more trivial case is to exclude the route from discovery on the dead/failed route. 

I think we should just drop those dead fsuckers rom the result if alive support enabled

Reply:
> I think we should just drop those dead suckers from the result if alive support enabled.

By just dropping the dead destination wouldn't it be to have the same result as with incorrect sorting for the "catch-all" rule? We drop the dead one and fall through to the "catch-all" rule anyway?
Would it be better to try to find alive destination for exactly same rule if multiple destination are configured? Otherwise - StatusBadGateway

Reply:
correct and this is why I've proposed UPD in the comment above. However this will be not an easy one, and a much simpler solution could be [a check for dead flag here](https://github.com/umputun/reproxy/blob/a73a16c12bbd376859667cb42258206a595d6c92/app/discovery/discovery.go#L134)

i.e. smth like this:

```
if m.dead {
  return dest, m.MatchType, false
}
```